services/thumbnails/pkg/thumbnail/storage/filesystem.go

```diff
@@ -62,14 +62,29 @@ func (s FileSystem) Put(key string, img []byte) error {
 	}
 
 	if _, err := os.Stat(imgPath); os.IsNotExist(err) {
-		f, err := os.Create(imgPath)
+		f, err := os.CreateTemp(dir, "tmpthumb")
 		if err != nil {
-			return errors.Wrapf(err, "could not create file \"%s\"", key)
+			return errors.Wrapf(err, "could not create temporary file for \"%s\"", key)
 		}
-		defer f.Close()
 
-		if _, err = f.Write(img); err != nil {
-			return errors.Wrapf(err, "could not write to file \"%s\"", key)
+		_, writeErr := f.Write(img) // write the thumbnail in the temporary file
+		f.Close()                   // close the file regardless of the error
```

Comment:
Why?

Reply:
I'm pretty sure is mandatory to close the file. Usually it's done with the `defer`, but I think we have to close the file before the potential rename that happens afterwards, so we can't defer the `Close` call.

Reply:
AFAIK the rename can happen without the file being closed before. So the `defer f.Close()` should be fine.

Reply:
> I think that, if you rename the file without closing it, there could be data still in the buffers waiting to be written (either in the user-space or kernel-space), so the risk of sending a partial thumbnail is higher.

Yes, the `defer f.Close()` would only work together with the `f.Sync()`. But the more I read about the more I think we should really do the  `f.Sync()` call. If you really want to properly fix the race condition it's needed.

Reply:
so I guess the plan is
write -> close -> sync -> rename
no defer call

Reply:
I think sync needs to be before closing. And I would still use a defer close call. (Why not?)
createtemp -> defer close -> write -> sync -> rename -> (according to the lwn article linked in my other comment a fsync on the parent dir would be needed here)

Reply:
is it possible to remove the file if it hasn't been closed yet? If there was a problem writing the contents we'll have to remove the temporary file instead of renaming it, not sure how the defer close call will behave in this scenario. I'm afraid it might not be possible to remove the file because the file is still open (might be the case for windows hosts, although I'm not sure if we're supporting them)

Reply:
> is it possible to remove the file if it hasn't been closed yet?

Yes. At least on Linux/Unix filesystems. The filedescriptor will stay valid until it's closed, even after the remove. The remove (unlink) is again just an operation on the parent directory.

Let's not worry about Windows hosts for now (we're not supporting that currently), I think the semantics are really different there. It would probably need as different implementation.